### PR TITLE
pallet vesting / update to use fungible

### PR DESCRIPTION
**Pull request by gilescope**
This updates the vesting pallet to use the new fungible traits rather than the now deprecated `Currency` and `ReserveCurrency` traits ( https://github.com/paritytech/substrate/pull/12951 ). Partial fix of https://github.com/paritytech/polkadot-sdk/issues/226.

TODO:

- before fees/tips could be paid from unvested funds.
- `add_locks` in benchmark needs a rethink because RuntimeFreezeReasons isn't a `[u8;8]`
- migrations

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3854482

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3854893

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3858175

File: polkadot/runtime/common/src/claims.rs [hunk 1]
```diff
@@ -171,7 +176,14 @@ pub mod pallet {
...
 		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
-		type VestingSchedule: VestingSchedule<Self::AccountId, Moment = BlockNumberFor<Self>>;
+		type VestingSchedule: VestingSchedule<
+			Self::AccountId,
+			Moment = BlockNumberFor<Self>,
+			Fungible = Self::Fungible,
+		>;
+		type Fungible: fungible::Inspect<Self::AccountId, Balance = Self::Balance>
+			+ fungible::Mutate<Self::AccountId>;
+		type Balance: Balance + MaybeSerializeDeserialize;
```
Comment (kianenigma) [inline polkadot/runtime/common/src/claims.rs:186]: No no, `Balance` can be derived from `type Fungible`.

File: polkadot/runtime/common/src/claims.rs [hunk 2]
```diff
@@ -171,7 +176,14 @@ pub mod pallet {
 	pub trait Config: frame_system::Config {
 		/// The overarching event type.
 		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
-		type VestingSchedule: VestingSchedule<Self::AccountId, Moment = BlockNumberFor<Self>>;
+		type VestingSchedule: VestingSchedule<
+			Self::AccountId,
+			Moment = BlockNumberFor<Self>,
+			Fungible = Self::Fungible,
+		>;
+		type Fungible: fungible::Inspect<Self::AccountId, Balance = Self::Balance>
```
Comment (kianenigma) [inline polkadot/runtime/common/src/claims.rs:184]:
```suggestion
		type Currency: fungible::Inspect<Self::AccountId, Balance = Self::Balance>
```

Comment (kianenigma) [reply polkadot/runtime/common/src/claims.rs:184]: Or `type NativeToken`, as suggested in https://github.com/paritytech/polkadot-sdk/issues/226#issuecomment-1691692395

File: substrate/frame/support/src/traits/tokens/fungible/freeze.rs [hunk 3]
```diff
@@ -121,3 +121,47 @@ pub trait Mutate<AccountId>: Inspect<AccountId> {
 		Self::set_frozen(id, who, a, Fortitude::Polite)
 	}
 }
+
+/// A vesting schedule over a fungible. This allows a particular fungible to have vesting limits
+/// applied to it.
+pub trait VestingSchedule<AccountId>: Sized {
```
Comment (kianenigma) [inline substrate/frame/support/src/traits/tokens/fungible/freeze.rs:127]: why is this moved?

File: polkadot/runtime/common/src/claims.rs (hunk 1 above)
Comment (gilescope) [reply polkadot/runtime/common/src/claims.rs:186]: Agreed but the pallet needs `Balance` to be `MaybeSerializeDesirialize`. There alas isn't another way to enforce this bound without pulling `Balance` out into a separate associated type that I know of.

File: substrate/frame/support/src/traits/tokens/fungible/freeze.rs (hunk 3 above)
Comment (gilescope) [reply substrate/frame/support/src/traits/tokens/fungible/freeze.rs:127]: At a later date `Currency` is likely to be removed. At that point it's going to be all alone in the `traits::tokens::currency` namespace. Is there a better namespace for it to migrate to or should we leave it where it is for now even though it now takes a `Fungible`?

File: polkadot/runtime/common/src/claims.rs (hunk 1 above)
Comment (gilescope) [reply polkadot/runtime/common/src/claims.rs:186]: Ah yes, the trait bound on the pallet seems to work nicely. Perfect.

**Comment by JuaniRios** [comment]
Any update on this? https://github.com/paritytech/polkadot-sdk/pull/3951 is almost merged, and if we merge this I think we can finally have vesting of pallet-assets tokens on AssetHub, which would make our flow at Polimec way better.

**Comment by bkchr** [comment]
There was no update for more than a year. Stuff doesn't need to kept open indefinitely. Take the code and open a new pr.
